### PR TITLE
Implement `dyn Trait` syntax (RFC 2113)

### PR DESCRIPTION
**Pull request by petrochenkov**
cc https://github.com/rust-lang/rust/issues/44662
r? @nikomatsakis

### CODE REVIEW

File: src/test/run-pass/dyn-trait.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+#![feature(dyn_trait)]
+
+use std::fmt::Display;
+
+static BYTE: u8 = 33;
+
+fn main() {
+    let x: &(dyn 'static + Display) = &BYTE;
+    let y: Box<dyn Display + 'static> = Box::new(BYTE);
```
Comment (nikomatsakis) [inline src/test/run-pass/dyn-trait.rs]: can we get a test like `dyn Vec<T>` -- i.e., something not a trait?

**Review by nikomatsakis — APPROVED** [review]
r=me modulo new test

**Comment by aidanhs** [comment]
```
[00:40:50] failures:

[00:40:50] 

[00:40:50] ---- [ui] ui/issue-44406.rs stdout ----

[00:40:50] 	normalized stderr:

[00:40:50] error: expected identifier, found keyword `true`

[00:40:50]   --> $DIR/issue-44406.rs:18:10

[00:40:50]    |

[00:40:50] 18 |     foo!(true);

[00:40:50]    |          ^^^^

[00:40:50] 

[00:40:50] error: expected type, found keyword `true`

[00:40:50]   --> $DIR/issue-44406.rs:18:10

[00:40:50]    |

[00:40:50] 13 |         bar(baz: $rest)

[00:40:50]    |                - help: did you mean to use `;` here?

[00:40:50] ...

[00:40:50] 18 |     foo!(true);

[00:40:50]    |          ^^^^ expecting a type here because of type ascription

[00:40:50] 

[00:40:50] error: expected one of `!`, `&&`, `&`, `(`, `*`, `.`, `;`, `<`, `?`, `[`, `_`, `dyn`, `extern`, `fn`, `for`, `impl`, `unsafe`, `}`, an operator, or lifetime, found `true`

[00:40:50]   --> $DIR/issue-44406.rs:18:10

[00:40:50]    |

[00:40:50] 13 |         bar(baz: $rest)

[00:40:50]    |                 - expected one of 20 possible tokens here

[00:40:50] ...

[00:40:50] 18 |     foo!(true);

[00:40:50]    |          ^^^^ unexpected token

[00:40:50] 

[00:40:50] error: aborting due to 3 previous errors
```
@bors r-

**Pull request merged**